File: src/main/scala/it/unibo/agar/view/LocalView.scala
package it.unibo.agar.view

import akka.actor.typed.ActorRef
import it.unibo.agar.model.DistributedGameStateManager
import it.unibo.agar.model.PlayerActor.{PlayerActorMessage, Terminate}

import java.awt.Graphics2D
import javax.swing.SwingUtilities
import scala.swing.*
import scala.swing.event.*

class LocalView(manager: DistributedGameStateManager, playerId: String, playerActorRef: ActorRef[PlayerActorMessage]) extends MainFrame:

  title = s"Agar.io - Local View ($playerId)"
  preferredSize = new Dimension(800, 800)

  override def closeOperation(): Unit = {} // Disabilita la chiusura automatica

  contents = new Panel:
    listenTo(keys, mouse.moves)
    focusable = true
    requestFocusInWindow()

    override def paintComponent(g: Graphics2D): Unit =
      val world = manager.world
      val playerOpt = world.players.find(_.id == playerId)
      val (offsetX, offsetY) = playerOpt
        .map(p => (p.x - size.width / 2.0, p.y - size.height / 2.0))
        .getOrElse((0.0, 0.0))
      AgarViewUtils.drawWorld(g, world, offsetX, offsetY)

    reactions += { case e: event.MouseMoved =>
      val mousePos = e.point
      val playerOpt = manager.world.players.find(_.id == playerId)
      playerOpt.foreach: player =>
        val dx = (mousePos.x - size.width / 2) * 0.01
        val dy = (mousePos.y - size.height / 2) * 0.01
        manager.movePlayerDirection(dx, dy)
      repaint()
    }

  listenTo(this)

  reactions += {
    case WindowClosing(_) =>
      playerActorRef ! Terminate()
  }

  def showView(): Unit =
    SwingUtilities.invokeLater(() => {
      println("showing")
      centerOnScreen()
      open()
      peer.toFront()
      peer.requestFocusInWindow()
    })

  def closeView(): Unit =
    SwingUtilities.invokeLater(() => {
      println("closing")
      close()
      dispose()
    })

  def showMessage(msg: String): Unit =
    SwingUtilities.invokeLater(() => {
      javax.swing.JOptionPane.showMessageDialog(peer, msg, "Game Ended", javax.swing.JOptionPane.INFORMATION_MESSAGE)
    })

File: src/main/java/it/unibo/agar/view/LocalView.java
package it.unibo.agar.view;

import it.unibo.agar.model.*;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.rmi.RemoteException;
import java.util.*;

public class LocalView extends JFrame {

    private static final double SENSITIVITY = 2;
    private final GamePanel gamePanel;
    private final DistributedGameStateManager gameStateManager;
    private final String playerId;
    private final PlayerController playerControllerStub;

    public LocalView(DistributedGameStateManager gameStateManager, String playerId, PlayerController playerControllerStub) {
        this.gameStateManager = gameStateManager;
        this.playerId = playerId;
        this.playerControllerStub = playerControllerStub;

        setTitle("Agar.io - Local View (" + playerId + ")");
        setPreferredSize(new Dimension(600, 600));
        setDefaultCloseOperation(EXIT_ON_CLOSE);

        gamePanel = new GamePanel(gameStateManager, playerId);
        add(gamePanel, BorderLayout.CENTER);

        addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                try {
                    playerControllerStub.terminate(true);
                } catch (RemoteException ex) {
                    throw new RuntimeException(ex);
                }
            }
        });

        setupMouseControls();
        pack();
        setLocationRelativeTo(null); // Center on screen
        setFocusable(true);
        requestFocusInWindow();
        setVisible(true);
    }

    private void setupMouseControls() {
        gamePanel.addMouseMotionListener(new MouseMotionAdapter() {
            @Override
            public void mouseMoved(MouseEvent e) {
                Optional<Player> playerOpt = gameStateManager.getWorld().getPlayerById(playerId);
                if (playerOpt.isPresent()) {
                    Point mousePos = e.getPoint();
                    // Player is always in the center of the local view
                    double viewCenterX = gamePanel.getWidth() / 2.0;
                    double viewCenterY = gamePanel.getHeight() / 2.0;

                    double dx = mousePos.x - viewCenterX;
                    double dy = mousePos.y - viewCenterY;

                    // Normalize the direction vector
                    double magnitude = Math.hypot(dx, dy);
                    if (magnitude > 0) { // Avoid division by zero if mouse is exactly at center
                        gameStateManager.movePlayerDirection((dx / magnitude) * SENSITIVITY, (dy / magnitude) * SENSITIVITY);
                    } else {
                        gameStateManager.movePlayerDirection(0, 0); // Stop if mouse is at center
                    }
                    // Repainting is handled by the main game loop timer
                }
            }
        });
    }

    public void repaintView() {
        if (gamePanel != null) {
            gamePanel.repaint();
        }
    }

    public void showView() {
        SwingUtilities.invokeLater(() -> {
            setLocationRelativeTo(null);
            setVisible(true);
            toFront();
            requestFocusInWindow();
        });
    }

    public void closeView() {
        SwingUtilities.invokeLater(this::dispose);
    }

    public void showMessage(String msg) {
        SwingUtilities.invokeLater(() ->
                JOptionPane.showMessageDialog(this, msg, "Game Ended", JOptionPane.INFORMATION_MESSAGE)
        );
    }
}
